otherlibs/cram/bin/run.ml

```diff
@@ -46,133 +56,264 @@ let run_expect_test file ~f =
+  in
+  List.iter cram_to_output ~f:(fun block ->
+      match (block : _ Cram_lexer.block) with
+      | Comment lines -> List.iter lines ~f:add_line
+      | Command
+          ( { command; output_file = _ }
+          , { exit_code; build_path_prefix_map = _ }
+          , output ) -> (
+        List.iteri command ~f:(fun i line ->
+            let line =
+              sprintf "%c %s"
+                ( if i = 0 then
+                  '$'
+                else
+                  '>' )
+                line
+            in
+            add_line_prefixed_with_two_space line);
+        String.split_lines output
+        |> List.iter ~f:add_line_prefixed_with_two_space;
+        match exit_code with
+        | 0 -> ()
+        | n -> add_line_prefixed_with_two_space (sprintf "[%d]" n) ));
+  Buffer.contents buf
+
+let create_sh_script cram_stanzas ~temp_dir : sh_script =
+  let script = Path.relative temp_dir "main.sh" in
+  let oc = Io.open_out ~binary:true script in
+  let script = Path.to_string script in
+  let file name = Path.relative temp_dir name |> Path.to_absolute_filename in
+  let sh_path path = quote_for_sh (translate_path_for_sh path) in
+  let metadata_file = file "cram.metadata" in
+  let metadata_file_sh_path = sh_path metadata_file in
+  let loop i block =
+    let i = succ i in
+    match (block : _ Cram_lexer.block) with
+    | Comment _ as comment -> comment
+    | Command lines ->
+      let file ~ext = file (sprintf "%d%s" i ext) in
+      (* Shell code written by the user might not be properly terminated. For
+         instance the user might forgot to write [EOF] after a [cat <<EOF]. If
+         we wrote this shell code directly in the main script, it would hide the
+         rest of the script. So instead, we dump each user written shell phrase
+         into a file and then source it in the main script. *)
+      let user_shell_code_file = file ~ext:".sh" in
+      let user_shell_code_file_sh_path = sh_path user_shell_code_file in
+      cat_eof oc lines ~dest:user_shell_code_file;
+      (* Where we store the output of shell code written by the user *)
+      let user_shell_code_output_file = file ~ext:".output" in
+      let user_shell_code_output_file_sh_path =
+        sh_path user_shell_code_output_file
       in
-      translate_path_for_sh prog |> quote_for_sh
+      fprln oc ". %s > %s 2>&1" user_shell_code_file_sh_path
+        user_shell_code_output_file_sh_path;
+      fprln oc {|printf "$?\0$%s\0" >> %s|} _BUILD_PATH_PREFIX_MAP
+        metadata_file_sh_path;
+      Command { command = lines; output_file = user_shell_code_output_file }
+  in
+  let cram_to_output = List.mapi ~f:loop cram_stanzas in
+  close_out oc;
+  { script; cram_to_output; metadata_file }
+
+let display_with_bars s =
+  List.iter (String.split_lines s) ~f:(Printf.eprintf "| %s\n")
+
+let run ~sanitizer ~file lexbuf =
+  let temp_dir =
+    let suffix = Filename.basename file in
+    Temp.dir ~prefix:".dune.cram" ~suffix
```

Comment:
Why the leading dot? I think it's quite unusual to do that

Reply:
Oops, that was supposed to be a trailing `.`. To make it easier to tell apart from the suffix.